Tests/JSSpecCompiler/test-runner.cpp

```diff
@@ -0,0 +1,184 @@
+            if (!dump_ast_option.is_empty())
+                dump_ast_option.append(","sv);
+            dump_ast_option.append(flag.name);
+        }
+    }
+    if (!dump_ast_option.is_empty())
+        result.append(format_into_byte_buffer("--dump-ast={}", dump_ast_option.string_view()));
+
+    if (test_source.has_extension(".cpp"sv))
+        result.append(format_into_byte_buffer("-xc++"sv));
+
+    result.append(test_source.string().to_byte_buffer());
+
+    return result;
+}
+
+ErrorOr<ByteBuffer> read(LexicalPath const& path)
+{
+    auto file = TRY(Core::File::open(path.string(), Core::File::OpenMode::Read));
+    return MUST(file->read_until_eof());
+}
+
+void check_expectations(LexicalPath const& path_to_expectation, LexicalPath const& path_to_captured_output, bool should_update_expectations)
+{
+    struct PathPair {
+        LexicalPath expectation;
+        LexicalPath result;
+    };
+    Vector<PathPair> file_pairs_to_check;
+
+    file_pairs_to_check.append({
+        .expectation = path_to_expectation,
+        .result = path_to_captured_output,
+    });
+
+    auto out = MUST(Core::File::standard_error());
+
+    for (auto const& [expectation_path, result_path] : file_pairs_to_check) {
+        auto result_content = read(result_path);
+
+        if (should_update_expectations && !result_content.is_error()) {
+            using namespace FileSystem;
+            MUST(copy_file_or_directory(expectation_path.string(), result_path.string(),
+                RecursionMode::Disallowed, LinkMode::Disallowed, AddDuplicateFileMarker::No));
+        }
+
+        auto expectation = read(expectation_path);
+
+        bool read_successfully = !(expectation.is_error() || result_content.is_error());
+        EXPECT(read_successfully);
+
+        if (read_successfully) {
+            bool are_equal = expectation.value() == result_content.value();
+            EXPECT(are_equal);
+
+            if (!are_equal) {
+                dbgln("Files {} and {} do not match!", expectation_path.string(), result_path.string());
+                auto maybe_diff = Diff::from_text(expectation.value(), result_content.value());
+                if (!maybe_diff.is_error()) {
+                    for (auto const& hunk : maybe_diff.value())
+                        MUST(Diff::write_unified(hunk, *out, Diff::ColorOutput::Yes));
+                }
+            }
+        }
+    }
+}
+
+TEST_CASE(test_regression)
+{
+    auto* source_dir_env = getenv("SERENITY_SOURCE_DIR");
+    auto* binary_dir_env = getenv("LAGOM_BINARY_DIR");
+
+    DeprecatedString source_dir;
+    DeprecatedString binary_dir;
+
+    if (source_dir_env == nullptr || binary_dir_env == nullptr) {
+        warnln("SERENITY_SOURCE_DIR or LAGOM_BINARY_DIR is not set. Assuming . for SERENITY_SOURCE_DIR and ./Build/lagom/bin for LAGOM_BINARY_DIR.");
+        source_dir = MUST(Core::System::getcwd());
+        binary_dir = LexicalPath::join(source_dir, "Build/lagom/bin"sv).string();
```

Comment:
These assumptions are kinda bunk. We don't want to assume Build/Lagom/bin.

For `path_to_compiler_binary`, you want something more like

```
auto path_to_self = LexicalPath(MUST(Core::System::current_executable_path())).parent();
auto path_to_compiler_binary = LexicalPath::join(path_to_self, compiler_binary);
```

Then we can yeet LAGOM_BINARY_DIR, we don't need it. 

For SERENITY_SOURCE_DIR, we should get rid of that as well, and instead assume that the tests will be in $PWD/Tests or some such. then you can set the working directory of the test runner to Tests/JSSpecCompiler/ and reference them from Core::System::cwd() or whatever we called that.

I'm not a fan of these build system specifics leaking into the binaries. ideally they should be relocatable. If you want we could change the way TestSuite::main() calls m_setup to let it take the Main::Arguments to parse extra ones. or take the ArgsParser by reference before calling parse(). That way when run in CI or via the scripts, we can assume the "standard paths", and you can also pass in --test-dir or --compiler-path if you like to the binary to run it manually.

These steps here should be in TEST_SETUP() and stored in some file scope globals rather than in the TEST_CASE, imo.